src/AI/WallyAI.java

```diff
@@ -0,0 +1,788 @@
+        GameAction bestAttack = null;
+        double bestDamage = 0;
+        for( GameActionSet actionSet : actionSets )
+        {
+          // See if we have the option to attack.
+          if( actionSet.getSelected().getType() == GameAction.ActionType.ATTACK )
+          {
+            for( GameAction action : actionSet.getGameActions() )
+            {
+              Unit target = gameMap.getLocation(action.getTargetLocation()).getResident();
+              double damage = target.model.getCost() * Math.min(target.getPreciseHP(), CombatEngine.simulateBattleResults(unit, target, gameMap, unit.x, unit.y).defenderHPLoss);
+              if( damage > bestDamage )
+              {
+                bestDamage = damage;
+                bestAttack = action;
+              }
+            }
+          }
+        }
+        if( null != bestAttack )
+        {
+          log(String.format("%s is shooting %s", unit.toStringWithLocation(), gameMap.getLocation(bestAttack.getTargetLocation()).getResident()));
+          actions.offer(bestAttack);
+          stateChange = true;
+          break;
+        }
+        else // tempqueue the siege unit if it can't attack
+        {
+          tempQueue.offer(unit);
+        }
+      }
+
+      // reset our units back into unitQueue if we need to calculate more
+      if( actions.isEmpty() )
+      {
+        unitQueue.addAll(tempQueue);
+        tempQueue.clear();
+      }
+
+      // Try to get confirmed kills with mobile strikes.
+      if( actions.isEmpty() && !unitQueue.isEmpty() )
+      {
+        boolean foundKill = false;
+        // Get a count of enemy forces.
+        Map<Commander, ArrayList<Unit>> unitLists = AIUtils.getEnemyUnitsByCommander(myCo, gameMap);
+        for( Commander co : unitLists.keySet() )
+        {
+          // log(String.format("Hunting CO %s's units", co.coInfo.name));
+          if( myCo.isEnemy(co) )
+          {
+            for( Unit target : unitLists.get(co) )
+            {
+              // log(String.format("  Would like to kill: %s", target.toStringWithLocation()));
+              ArrayList<XYCoord> coordsToCheck = Utils.findLocationsInRange(gameMap, new XYCoord(target.x, target.y), 1, mobileAttackRange);
+              Map<XYCoord, Unit> neededAttacks = new HashMap<XYCoord, Unit>();
+              double damage = 0;
+
+              for( XYCoord xyc : coordsToCheck )
+              {
+                Location loc = gameMap.getLocation(xyc);
+                Unit resident = loc.getResident();
+                if( null != resident && resident.CO == myCo && !resident.isTurnOver &&
+                    resident.canAttack(target.model, xyc.getDistance(target.x, target.y), false))
+                {
+                  damage += CombatEngine.simulateBattleResults(resident, target, gameMap, xyc.xCoord, xyc.yCoord).defenderHPLoss;
+                  neededAttacks.put(xyc, resident);
+                  if( damage >= target.getPreciseHP() )
+                    break;
```

Comment:
Couldn't you also set `foundKill` to true here and avoid going into `findAssaultKills`?

Reply:
I guess that's the first condition you check in `findAssaultKills`, so it works out, but still.

Reply:
I didn't think of it.